### PR TITLE
Ensure Redis and SqlServer cache operations silently fail

### LINKED ISSUE #193: More reliable cache operations
As part of discussions while working on https://github.com/aspnet/Session/issues/99, there were proposals for adding logic in caching to make database operations more reliable. For example, sql database operations may fail intermittently. In this case it may be a good idea to add retry logic and allow users to specify retry policies. This probably needs some more design work.

cc @mikeharder @Tratcher @davidfowl

### PR DESCRIPTION
**Pull request by natemcmaster**
Resolve #193 

Per our design discussion, this adds top-level catches for all exceptions with optional logging for suppressed exceptions.

NB: introduces dependency on logging abstractions.

cc @JunTaoLuo @Tratcher @pakrym

### CODE REVIEW

File: src/Microsoft.Extensions.Caching.Redis/RedisCache.cs [hunk 1]
```diff
@@ -380,6 +426,14 @@ private void Refresh(string key, DateTimeOffset? absExpr, TimeSpan? sldExpr)
             return absoluteExpiration;
         }
 
+        private void LogSuppressedException(Exception ex)
+        {
+            if (_logger != null)
+            {
+                _logger.LogDebug("An exception during cache operation was suppressed: {message}", ex.ToString());
```
Comment (Tratcher) [inline src/Microsoft.Extensions.Caching.Redis/RedisCache.cs]: you don't need to insert the message, you can pass down the entire exception object.

File: src/Microsoft.Extensions.Caching.Redis/RedisCache.cs [hunk 2]
```diff
@@ -380,6 +426,14 @@ private void Refresh(string key, DateTimeOffset? absExpr, TimeSpan? sldExpr)
             return absoluteExpiration;
         }
 
+        private void LogSuppressedException(Exception ex)
+        {
+            if (_logger != null)
+            {
+                _logger.ExceptionSuppressed(ex);
```
Comment (pakrym) [inline src/Microsoft.Extensions.Caching.Redis/RedisCache.cs]: optional nit: you can do `_logger?.ExceptionSuppressed(ex);` in catch blocks

**Comment by pakrym** [comment]
:shipit:

**Comment by JunTaoLuo** [comment]
Would be nice to have a few tests to make sure exceptions are caught and the correct messages are logged.

**Comment by natemcmaster** [comment]
Reverted in 7abe1ccc0aab2f43066e82f3ba5bff2c40ea41d8

**Pull request merged**